File: content/thoughts/university/twenty-three-twenty-four/sfwr-3bb4/a2/average.py
from __future__ import annotations
from typing import Any
from threading import Thread
import random, time

class Worker(Thread):
  def __init__(self, arr: list[Any], start: int, end: int) -> None:
    super().__init__()
    self._arr, self._start, self._end = arr, start, end
    self._result = 0.0

  def run(self) -> None:
    for i in range(self._start, self._end): self._result += self._arr[i]

def sequentialaverage(a: int): return sum(a) / len(a)

def parallelaverage(a: int, p: int):
  n = len(a)
  chunk_size = n // p
  workers: list[Worker]= []

  for i in range(p):
    start = i * chunk_size
    end = (i + 1) * chunk_size if i < p - 1 else n
    worker = Worker(a, start, end)
    workers.append(worker)
    worker.start()

  for worker in workers: worker.join()

  total_sum = 0.0
  for worker in workers: total_sum += worker._result

  return total_sum / n

def average(n, p = 1):
  a = [random.randint(0, 1000) for i in range(n)]

  start = time.time_ns() / 1000
  avg = sequentialaverage(a)
  end  = time.time_ns() / 1000
  print("Sequential:", avg, "Time:", end - start, "µs")

  start = time.time_ns() / 1000
  avg = parallelaverage(a, p)
  end  = time.time_ns() / 1000
  print("Parallel:", avg, "Time:", end - start, "µs")

File: content/thoughts/university/twenty-three-twenty-four/sfwr-3bb4/a2/Average.java
import java.util.Random;

class Worker extends Thread {
  private int[] arr;
  private int start, end;
  private double result;

  public Worker(int[] arr, int start, int end) {
      this.arr = arr;
      this.start = start;
      this.end = end;
      this.result = 0.0;
  }

  @Override
  public void run() {
      for (int i = start; i < end; i++) {
          result += arr[i];
      }
  }

  public double getResult() {
      return result;
  }
}

public class Average {
    static double sequentialaverage(int a[]) {
        // a.length > 0
        double s = 0;
        for (int i = 0; i < a.length; i++) s += a[i];
        return s / a.length;
    }
    static double parallelaverage(int a[]) {
        // a.length > 0
        int n = a.length;
        int mid = n / 2;

        Worker worker1 = new Worker(a, 0, mid);
        Worker worker2 = new Worker(a, mid, n);

        worker1.start();
        worker2.start();

        try {
            worker1.join();
            worker2.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        double sum1 = worker1.getResult();
        double sum2 = worker2.getResult();

        return (sum1 + sum2) / n;
    }
    public static void main(String args[]) {
        int n = Integer.parseInt(args[0]); // compute the average of n random numbers
        int[] a = new int[n];
        Random rand = new Random();
        for (int i = 0; i < n; i++) a[i] = rand.nextInt(10000);
        
        long start = System.currentTimeMillis();
        double avg = sequentialaverage(a);
        long end = System.currentTimeMillis();
        System.out.println("Sequential: " + avg + " Time: " + (end - start) + " ms");

        start = System.currentTimeMillis();
        avg = parallelaverage(a);
        end = System.currentTimeMillis();
        System.out.println("Parallel: " + avg + " Time: " + (end - start) + " ms");
    }
}
